### PR TITLE
CDPSDX-3900 Add Logging to DB Backup/Restore for Errors where Wrong N…

### PR DESCRIPTION
**Pull request by grace303303**
…ode is Used

*Restore will be updated after we determine we want to make the logging this detailed. 

**JIRA**: JIRA: [CDPSDX-3900](https://jira.cloudera.com/browse/CDPSDX-3900)
**ISSUE**: When a user ran the salt command to manually backup database on a wrong node the backup fails because it is missing values in the pillar file or in postgres.

**SOLUTION**: To make the logging more precise, I added in options when running the bash command and use `getopts` to get the argument of each option.

**BEFORE** (the log):
image

**AFTER** (the log):
Tested with a successful backup:

Tested with a successful backup with database names defined:
Screen Shot 2023-02-21 at 3 45 16 PM

Tested with backup on the wrong node:
Screen Shot 2023-02-23 at 9 05 11 PM

Tested with a successful restore (the restore log has some decoding issues which we may need to fix in the future):
Screen Shot 2023-02-21 at 9 18 28 PM

Tested with restore on the wrong node:
Screen Shot 2023-02-23 at 9 06 50 PM

Also tested by setting up configure_remote_db to be "None" so it will call another command in the jinjia file.

### CODE REVIEW

File: orchestrator-salt/src/main/resources/salt/salt/postgresql/disaster_recovery/scripts/backup_db.sh [hunk 1]
```diff
@@ -27,26 +27,73 @@ doLog() {
...
+        STORAGE_URL="$OPTARG"
+        doLog "Object Storage Service url to place backups: $STORAGE_URL"
+        ;;
+    h  )
+        HOST="$OPTARG"
+        doLog "PostgreSQL host name: $HOST"
+        ;;
+    p  )
+        PORT="$OPTARG"
+        doLog "PostgreSQL port: $HOSTNAME"
```
Comment (votez) [inline orchestrator-salt/src/main/resources/salt/salt/postgresql/disaster_recovery/scripts/backup_db.sh:48]: $PORT

File: orchestrator-salt/src/main/resources/salt/salt/postgresql/disaster_recovery/scripts/backup_db.sh [hunk 2]
```diff
@@ -27,26 +27,73 @@ doLog() {
...
+
+BACKUP_LOCATION="$STORAGE_URL"
+HOST="$HOST"
+PORT="$PORT"
+USERNAME="$USERNAME"
+RANGERGROUP="$GROUP"
+CLOSECONNECTIONS="$CLOSE_CONNECT"
+# Assign values to db name based on if var DB_NAME has been set.
+if [ -z ${DB_NAME+x} ]; then
+    DATABASENAMES=""
```
Comment (votez) [inline orchestrator-salt/src/main/resources/salt/salt/postgresql/disaster_recovery/scripts/backup_db.sh:93]: please `doLog` this situation - we better know databases were not listed

File: orchestrator-salt/src/main/resources/salt/salt/postgresql/disaster_recovery/scripts/backup_db.sh (hunk 1 above)
Comment (grace303303) [reply orchestrator-salt/src/main/resources/salt/salt/postgresql/disaster_recovery/scripts/backup_db.sh]: Thank you!! Fixed!

File: orchestrator-salt/src/main/resources/salt/salt/postgresql/disaster_recovery/scripts/backup_db.sh (hunk 2 above)
Comment (grace303303) [reply orchestrator-salt/src/main/resources/salt/salt/postgresql/disaster_recovery/scripts/backup_db.sh]: Thank you @votez added logs for that!

File: orchestrator-salt/src/main/resources/salt/salt/postgresql/disaster_recovery/scripts/backup_db.sh [hunk 3]
```diff
@@ -27,26 +27,71 @@ doLog() {
...
-  doLog "  2. PostgreSQL host name."
-  doLog "  3. PostgreSQL port."
-  doLog "  4. PostgreSQL user name."
-  doLog "  5. Ranger admin group."
-  doLog "  6. Whether or not to close connections for the database while it is being backed up."
-  doLog "  7-10. (optional) Names of the databases to backup. If not given, will backup ranger and hive databases."
-  exit 1
-fi
+log_invalid_input_reason() {
+  doLog "This might be caused by the command is not running on the master Getaway node or there has never had a backup/restore from the CDP CLI run."
```
Comment (sxxgrc) [inline orchestrator-salt/src/main/resources/salt/salt/postgresql/disaster_recovery/scripts/backup_db.sh:31]: Grammar fix: `"This might be caused by the command not being run on the Primary Gateway node or due to never having run a backup/restore via the CDP CLI before."`

File: orchestrator-salt/src/main/resources/salt/salt/postgresql/disaster_recovery/scripts/backup_db.sh [hunk 4]
```diff
@@ -27,26 +27,71 @@ doLog() {
...
 
-BACKUP_LOCATION="$1"
-HOST="$2"
-PORT="$3"
-USERNAME="$4"
-RANGERGROUP="$5"
-CLOSECONNECTIONS="$6"
-DATABASENAMES="${@: 7}"
+# Print out all required inputs if they are passed in.
+while getopts ":s:h:p:u:g:c:n:" OPTION; do
```
Comment (sxxgrc) [inline orchestrator-salt/src/main/resources/salt/salt/postgresql/disaster_recovery/scripts/backup_db.sh:35]: I understand why you're changing this to not use positional arguments, but please note that this will break backwards compatibility in case any customers have their own scripts for running backup/restore. Please confirm with Kalyan that this is okay.

Please also make sure that running a backup/restore via the CB flow is still functional. You can check this by running a resize and making sure the backup/restore steps are successful.

File: orchestrator-salt/src/main/resources/salt/salt/postgresql/disaster_recovery/scripts/backup_db.sh [hunk 5]
```diff
@@ -27,26 +27,71 @@ doLog() {
...
+    c  )
+        CLOSE_CONNECT="$OPTARG"
+        doLog "Whether or not to close connections for the database while it is being backed up: $CLOSE_CONNECT"
+        ;;
+    n  )
+        DB_NAME+=("$OPTARG")
+        DB_NAME_ITEM="$OPTARG"
+        doLog "Names of the databases to backup. If not given, will backup ranger and hive databases: $DB_NAME_ITEM"
+        ;;
+    \? ) doLog "Unknown option: -$OPTARG" >&2;
```
Comment (sxxgrc) [inline orchestrator-salt/src/main/resources/salt/salt/postgresql/disaster_recovery/scripts/backup_db.sh:66]: Is the `>&2` useful here?

File: orchestrator-salt/src/main/resources/salt/salt/postgresql/disaster_recovery/scripts/backup_db.sh [hunk 6]
```diff
@@ -27,26 +27,71 @@ doLog() {
...
+        exit 1;;
+    :  ) echo "Missing option argument for -$OPTARG" >&2;
```
Comment (sxxgrc) [inline orchestrator-salt/src/main/resources/salt/salt/postgresql/disaster_recovery/scripts/backup_db.sh:68]: Why echo instead of log?

File: orchestrator-salt/src/main/resources/salt/salt/postgresql/disaster_recovery/scripts/backup_db.sh [hunk 7]
```diff
@@ -27,26 +27,71 @@ doLog() {
...
+        exit 1;;
+    esac
+done
+
+# If any of the required inputs are missing or the provided values are NULL, the script can not run successfully.
+[ -z ${STORAGE_URL+x} ] || [ -z ${GROUP+x} ] || [ -z ${CLOSE_CONNECT+x} ]||
+[ -z "$STORAGE_URL" ] || [ -z "$GROUP" ] || [ -z "$CLOSE_CONNECT" ] &&
+doLog "At least one mandatory parameter is not set in disaster_recovery.sls: Object Storage Service URL; Ranger admin group; close connections or not during backup." && log_invalid_input_reason && exit 1
+[ -z ${HOST+x} ] || [ -z ${PORT+x}  ] || [ -z ${USERNAME+x} ] ||
+[ -z "$HOST" ] || [ -z "$PORT" ] || [ -z "$USERNAME" ] &&
```
Comment (sxxgrc) [inline orchestrator-salt/src/main/resources/salt/salt/postgresql/disaster_recovery/scripts/backup_db.sh:79]: In `backup.sls` we see that the script can be called with 

```
/opt/salt/scripts/backup_db.sh{% if storage %} -s {{storage}}{% endif %} -h "" -p "" -u ""{% if group %} -g {{group}}{% endif %}{% if dbname %}{% for item in dbname_list %} -n {{item}}{% endfor %}{% endif %}
```
So there is a case where we run the script with `-h "" -p "" -u ""` so this would prevent backup/restore in that case. 

However, this case is only for when an external DB is not used which isn't the case for any customer, only some internal uses. I recommend confirming with Kalyan that we still want to support this case.

File: orchestrator-salt/src/main/resources/salt/salt/postgresql/disaster_recovery/scripts/backup_db.sh [hunk 8]
```diff
@@ -27,26 +27,71 @@ doLog() {
...
+
+BACKUP_LOCATION="$STORAGE_URL"
+HOST="$HOST"
+PORT="$PORT"
+USERNAME="$USERNAME"
+RANGERGROUP="$GROUP"
+CLOSECONNECTIONS="$CLOSE_CONNECT"
+# Assign values to db name based on if var DB_NAME has been set and if the value set to the var is null.
+if [ -z ${DB_NAME+x} ] || [ -z "$DB_NAME" ]; then
+  doLog "No database names are provided in disaster_recovery.sls. Will backup ranger and hive by default."
```
Comment (sxxgrc) [inline orchestrator-salt/src/main/resources/salt/salt/postgresql/disaster_recovery/scripts/backup_db.sh:90]: Below we already log:

```
if [[ -z "$DATABASENAMES" ]]; then
    doLog "INFO No database name provided. Will backup hive, ranger, profiler_agent and profiler_metric databases."
```

So this is not needed, also it's clear that we are, by default, backing up the profiler DBs as well, please update the line 64 log to make this clear.

File: orchestrator-salt/src/main/resources/salt/salt/postgresql/disaster_recovery/scripts/restore_db.sh [hunk 9]
```diff
@@ -25,24 +25,65 @@ doLog() {
   echo "$(date "+%Y-%m-%dT%H:%M:%SZ") $type_of_msg ""$msg" >>$LOGFILE
 }
 
-if [[ $# -lt 5 || $# -gt 6 || "$1" == "None" ]]; then
```
Comment (sxxgrc) [inline orchestrator-salt/src/main/resources/salt/salt/postgresql/disaster_recovery/scripts/restore_db.sh:28]: Please make the same corrections I defined for the backup_db.sh script.

**Review by sxxgrc — CHANGES_REQUESTED** [review]
I've left some comments with changes to logging and also concerns about some of the changes.

File: orchestrator-salt/src/main/resources/salt/salt/postgresql/disaster_recovery/scripts/backup_db.sh (hunk 3 above)
Comment (grace303303) [reply orchestrator-salt/src/main/resources/salt/salt/postgresql/disaster_recovery/scripts/backup_db.sh:31]: Updated this, thank you so much!

File: orchestrator-salt/src/main/resources/salt/salt/postgresql/disaster_recovery/scripts/backup_db.sh (hunk 4 above)
Comment (grace303303) [reply orchestrator-salt/src/main/resources/salt/salt/postgresql/disaster_recovery/scripts/backup_db.sh:35]: @sxxgrc That's a really good point - I didn't realize customers will run with their own scripts. Let me ask Kalyan about it. Thank you! I will update this thread after getting confirmation and testing with resize. If we do need to think of customers who use their own scripts, I will switch back to the first version PR which only has a few changes. https://github.com/hortonworks/cloudbreak/pull/14308/files

File: orchestrator-salt/src/main/resources/salt/salt/postgresql/disaster_recovery/scripts/backup_db.sh (hunk 5 above)
Comment (grace303303) [reply orchestrator-salt/src/main/resources/salt/salt/postgresql/disaster_recovery/scripts/backup_db.sh]: We have that in log so I guess I can remove it. Thank you! Removed.

File: orchestrator-salt/src/main/resources/salt/salt/postgresql/disaster_recovery/scripts/backup_db.sh (hunk 7 above)
Comment (grace303303) [reply orchestrator-salt/src/main/resources/salt/salt/postgresql/disaster_recovery/scripts/backup_db.sh:79]: @sxxgrc Sure! I will ask Kalyan about it. The added `-h` `-p` shouldn't impact how it functions though since the passed in paramaters are still empty. I tested this and it can not work, which is expected. When I tested it I don't know what is the purpose of this cmd though. Thank you for letting me know!

File: orchestrator-salt/src/main/resources/salt/salt/postgresql/disaster_recovery/scripts/backup_db.sh (hunk 8 above)
Comment (grace303303) [reply orchestrator-salt/src/main/resources/salt/salt/postgresql/disaster_recovery/scripts/backup_db.sh:90]: Updated! I removed the last sentence. Thank you!!

File: orchestrator-salt/src/main/resources/salt/salt/postgresql/disaster_recovery/scripts/restore_db.sh (hunk 9 above)
Comment (grace303303) [reply orchestrator-salt/src/main/resources/salt/salt/postgresql/disaster_recovery/scripts/restore_db.sh:28]: Updated those for restore as well. Thank you!

File: orchestrator-salt/src/main/resources/salt/salt/postgresql/disaster_recovery/scripts/backup_db.sh (hunk 6 above)
Comment (grace303303) [reply orchestrator-salt/src/main/resources/salt/salt/postgresql/disaster_recovery/scripts/backup_db.sh]: Thank you for noticing this! Updated.

**Comment by grace303303** [comment]
Hi team, sorry for the back and forth. After experiencing last Friday's escalation I realize updating the backup.sls file by changing the arguments is not a good idea - as what Santiago said, we need to inform everyone who is manually running the backup_db.sh file about this change if so because how they put in the inputs will be changed. I don't think it's worth the risk to make these many changes just to do error logging. So I'm closing this PR and will switch to use the first PR which is a much simpler change. CC @sxxgrc @kkalvagadda1 @votez
